### PR TITLE
Initial implementation of the multi-arch builder controller

### PR DESCRIPTION
**Pull request by droslean**
- Introduces the `MultiArchBuildConfig` Custom Resource.

example:
```yaml
apiVersion: ci.openshift.io/v1
kind: MultiArchBuildConfig
metadata:
  name: test-mabc
  namespace: ci
spec:
  build_spec:
    output:
      to:
        kind: ImageStreamTag
        name: managed-clonerefs:latest-test
        namespace: ci
    resources:
      requests:
        cpu: 100m
        memory: 200Mi
    runPolicy: Serial
    source:
      dockerfile: |-
        FROM gcr.io/k8s-prow/git:latest
        ADD ./clonerefs /clonerefs
      images:
      - from:
          kind: ImageStreamTag
          name: clonerefs:latest
        paths:
        - destinationDir: .
          sourcePath: /ko-app/clonerefs
      type: Dockerfile
    strategy:
      dockerStrategy:
        imageOptimizationPolicy: SkipLayers
        noCache: true
      type: Docker

```

- Implements the `multi-arch-builder-controller`
The controller reconciles the `MultiArchBuildConfig` and generates multiple builds one for each architecture that exists on the cluster. Once all builds succeed, the controller uses the `manifest-tool` binary to create a new image based on the output configuration that includes the manifest list with all images that have been built per architecture correspondingly.

/cc @openshift/test-platform 
/cc @bbguimaraes @jmguzik @danilo-gemoli @deepsm007  

Signed-off-by: Nikolaos Moraitis <nmoraiti@redhat.com>

### CODE REVIEW

File: pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+)
+
+const (
+	controllerName        = "multiarchbuildconfig"
+	nodeArchitectureLabel = "kubernetes.io/arch"
+	noReasonCondition     = "NoReason"
+
+	// TODO: Now we hardcode the URL for the heterogeneous cluster.
+	// TODO: We need to push it to the app.ci registry or quay.io.
+	registryURL = "registry.multi-build01.arm-build.devcluster.openshift.com"
```
Comment (droslean) [inline pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go:41]: @hongkailiu Can we start pushing the images to quay.io or we should push them to `app.ci` registry in order to make sure that the test-image-distributor will sync the image across all clusters?

File: cmd/multi-arch-builder-controller/main.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+	if err := mgr.Start(ctx); err != nil {
+		logrus.WithError(err).Fatal("Manager ended with error")
+	}
+
+	logrus.Info("Process ended gracefully")
+}
+
+func resolveNodeArchitectures(ctx context.Context, client coreclientset.NodeInterface) ([]string, error) {
+	ret := sets.New[string]()
+	nodeList, err := client.List(ctx, metav1.ListOptions{})
```
Comment (droslean) [inline cmd/multi-arch-builder-controller/main.go:101]: @deepsm007 This approach won't ensure to include an architecture from a scaled-down to 0 machineset. When we will start adding more architectures to the heterogeneous cluster, we will need to make sure that at least one node is running or change this approach to gather the architectures from the machinessets which is really hard now.

Comment (deepsm007) [reply cmd/multi-arch-builder-controller/main.go:101]: We need to add that in comment inside the code and keep track of it when we move from more than 2 archs

File: pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+	w, err := r.buildClient.Builds(mabc.Namespace).Watch(ctx, metav1.ListOptions{
+		LabelSelector: labels.Set{v1.MultiArchBuildConfigNameLabel: mabc.Name}.String(),
+	})
+	if err != nil {
+		return err
+	}
+
+Loop:
+	for {
+		event, ok := <-w.ResultChan()
```
Comment (droslean) [inline pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go:163]: @bbguimaraes Do you think that it's better to create another controller that watches/reconciles the Builds instead of running a watcher here?

File: pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+
+	cmdOutput := &bytes.Buffer{}
+	cmdError := &bytes.Buffer{}
+	cmd.Stdout = cmdOutput
+	cmd.Stderr = cmdError
+
+	logrus.Infof("Running command: %s", cmd.String())
+	err := cmd.Run()
+	if err != nil {
+		logrus.WithError(err).WithField("output", cmdOutput.String()).WithField("error_output", cmdError.String()).Info("manifest-tool command failed...")
```
Comment (jmguzik) [inline pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go:355]:
```suggestion
		logrus.WithError(err).WithField("output", cmdOutput.String()).WithField("error_output", cmdError.String()).Info("manifest-tool command failed")
```

File: pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+		return err
+	}
+
+	logrus.WithField("output", cmdOutput.String()).Info("manifest-tool command succeeded...")
```
Comment (jmguzik) [inline pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go:359]:
```suggestion
	logrus.WithField("output", cmdOutput.String()).Info("manifest-tool command succeeded")
```
... suggests that there is something more to it

File: pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go (hunk 1 above)
Comment (jmguzik) [reply pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go:40]: Afaik @hongkailiu is syncing images to quay so you would need to treat still app.ci as the source of truth.

File: pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+	}
+
+	if err := r.handleMultiArchBuildConfig(ctx, mabc); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *reconciler) handleMultiArchBuildConfig(ctx context.Context, mabc *v1.MultiArchBuildConfig) error {
```
Comment (jmguzik) [inline pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go:122]: This function is not executed in any of tests.

File: pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+		}
+
+		if checkAllBuildsFinished(builds) {
+			r.logger.Info("All builds have finished...")
+			return nil
+		}
+	}
+}
+
+func (r *reconciler) createBuildsForArchitectures(ctx context.Context, mabc *v1.MultiArchBuildConfig) (map[string]*buildv1.Build, error) {
```
Comment (jmguzik) [inline pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go:201]: As a consequence, this one is also not covered.

**Review by jmguzik — APPROVED** [review]
I did a high level review of the code. Generally LGTM for me, found really small nits and left comments about some parts of functionality not covered in tests at all. As this is a new tool, not messing with implementation of any currently used tools we have I give a green stamp with hold if somebody else wants to check as well.
/hold

File: cmd/multi-arch-builder-controller/main.go [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+	if err := multiarchbuildconfigv1.AddToScheme(mgr.GetScheme()); err != nil {
+		logrus.WithError(err).Fatal("Failed to add multiarchbuildconfig to scheme")
+	}
+
+	if err := buildv1.AddToScheme(mgr.GetScheme()); err != nil {
+		logrus.WithError(err).Fatal("Failed to add multiarchbuildconfig to scheme")
+	}
+
+	nodeArchitectures, err := resolveNodeArchitectures(ctx, kubeClient.Nodes())
```
Comment (bbguimaraes) [inline cmd/multi-arch-builder-controller/main.go:83]: Do we want to do this only once at initialization time or also keep track of nodes as they enter and leave the cluster?

File: pkg/api/multiarchbuildconfig/v1/types.go [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+type MultiArchBuildConfigState string
+
+const (
+	// SuccessState means all builds were completed without error (exit 0)
+	SuccessState MultiArchBuildConfigState = "success"
+	// FailureState means that all builds were completed with errors (exit non-zero)
+	FailureState MultiArchBuildConfigState = "failure"
+)
```
Comment (bbguimaraes) [inline pkg/api/multiarchbuildconfig/v1/types.go:51]: What if none of these are true?

File: pkg/controller/multiarchbuildconfig/buildclient.go
Comment (bbguimaraes) [inline pkg/controller/multiarchbuildconfig/buildclient.go:1]: Why are these necessary?

File: pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go (hunk 7 above)
Comment (droslean) [reply pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go:201]: I will add tests for this as well.

File: pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go [hunk 10]
```diff
@@ -0,0 +1,361 @@
...
+func (r *reconciler) handleMultiArchBuildConfig(ctx context.Context, mabc *v1.MultiArchBuildConfig) error {
+	targetImageRef := fmt.Sprintf("%s/%s", mabc.Spec.BuildSpec.CommonSpec.Output.To.Namespace, mabc.Spec.BuildSpec.CommonSpec.Output.To.Name)
+	r.logger.WithField("target_image", targetImageRef).Info("Start handling builds to generate image...")
+
+	builds, err := r.createBuildsForArchitectures(ctx, mabc)
+	if err != nil {
+		return fmt.Errorf("couldn't create builds for architectures: %s: %w", strings.Join(r.architectures, ","), err)
+	}
+
+	if err := r.watchAndHandleBuildEvents(ctx, mabc, builds); err != nil {
```
Comment (bbguimaraes) [inline pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go:131]: This will only give any indication of the progress when all builds are done, which I don't think will be appreciated by people watching their test results

File: pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go (hunk 3 above)
Comment (bbguimaraes) [reply pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go:172]: Why not do it in a reconciliation loop?  See also the previous comment.

File: pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go [hunk 11]
```diff
@@ -0,0 +1,361 @@
...
+		}
+		return nil
+	})
+}
+
+func (r *reconciler) pruneBuilds(ctx context.Context, builds map[string]*buildv1.Build) error {
+	for _, build := range builds {
+		if err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			b := &buildv1.Build{}
+			if err := r.client.Get(ctx, types.NamespacedName{Namespace: build.Namespace, Name: build.Name}, b); err != nil {
```
Comment (bbguimaraes) [inline pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go:253]: Why is this `Get` necessary?

File: pkg/api/multiarchbuildconfig/v1/types.go (hunk 9 above)
Comment (droslean) [reply pkg/api/multiarchbuildconfig/v1/types.go:51]: Now we assume that the resources are not yet processed. I think we should also add another state that indicates that the resource is in a Processing state.

File: cmd/multi-arch-builder-controller/main.go (hunk 8 above)
Comment (droslean) [reply cmd/multi-arch-builder-controller/main.go:83]: Keeping track is a good idea but for now, let's do it at the initialization time. I am investigating also how to utilize the machine sets in order to keep track of the architectures in real-time. I am planning some follow-ups on that front.

File: pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go (hunk 10 above)
Comment (droslean) [reply pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go:131]: The error of each of the builds is being recorded in the resource conditions as well.

File: pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go (hunk 11 above)
Comment (droslean) [reply pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go:253]: To avoid conflicts. The resource could change in the meantime by another process that is not related to this controller. This is an idiomatic way to manage the resources in a kube/openshift cluster.

File: pkg/controller/multiarchbuildconfig/buildclient.go
Comment (droslean) [reply pkg/controller/multiarchbuildconfig/buildclient.go:1]: For testing mostly. I need to write an adapter for the build client to use it in the tests.

File: pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go (hunk 6 above)
Comment (droslean) [reply pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go:122]: Yes, but all of the main operations inside are tested.

File: pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go (hunk 3 above)
Comment (droslean) [reply pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go:172]: We can add the builds with the multiarchbuildconfig label to be owned by the manager as well. We could potentially trigger the reconciliation process or write another controller for them. That will make things little bit more complicated though. I will dig into it.

File: pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go (hunk 10 above)
Comment (bbguimaraes) [reply pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go:131]: But only after all of them are done, since the resource is only modified after this point.  This is why I think we should treat `Build`s the same way we treat `MultiArchBuildConfig`s.  It's similar to what the payload plugin does with `ProwJob`s.

File: pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go (hunk 11 above)
Comment (bbguimaraes) [reply pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go:253]: But what are we gaining if we're just doing an arbitrary `Get` right before deleting it?  What conflict does this prevent?

If the intent is to delete the specific version created previously, shouldn't we use the values from `builds` when deleting?  Or do it this way:

https://github.com/openshift/ci-tools/blob/2dee011d0b7a2aa21cdf6dce77775a3c9bf4ace2/pkg/util/pods.go#L91

Comment (bbguimaraes) [reply pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go:253]: In any case,`ci-operator` doesn't delete `Build`s when they are done, we just let the namespace expire (our TTL is very short these days).  This is very useful if someone needs to investigate when something goes wrong.

Comment (droslean) [reply pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go:253]: Yea, the whole get/delete thing was to avoid conflicts if the build was mutated in the window between the get and the delete. The delete in that case will fail. However, I would agree that we shouldn't clean them up. I will check your suggestions. Definitely, the users should be able to investigate their build if it fails.

File: pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go (hunk 10 above)
Comment (droslean) [reply pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go:131]: Yea, when I wrote payload plugin controller, I though that it would be better to use another controller to watch prowjobs. However, I was hoping that the builds in this case, should be more trivial operation, and get away with it with just a watcher. I would prefer the controller though. I will make the changes.

File: pkg/api/multiarchbuildconfig/v1/types.go [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+func UpdateMultiArchBuildConfig(ctx context.Context, logger *logrus.Entry, client ctrlruntimeclient.Client, namespacedName types.NamespacedName, mutateFn func(mabcToMutate *MultiArchBuildConfig)) error {
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		multiArchBuildConfig := &MultiArchBuildConfig{}
+		if err := client.Get(ctx, namespacedName, multiArchBuildConfig); err != nil {
+			return fmt.Errorf("failed to get the MultiArchBuildConfig: %w", err)
+		}
+
+		mutateFn(multiArchBuildConfig)
+
+		logger.WithField("namespace", namespacedName.Namespace).WithField("name", namespacedName.Name).Info("Updating MultiArchBuildConfig...")
```
Comment (jmguzik) [inline pkg/api/multiarchbuildconfig/v1/types.go:71]:
```suggestion
		logger.WithField("namespace", namespacedName.Namespace).WithField("name", namespacedName.Name).Info("Updated MultiArchBuildConfig")
```

**Review by jmguzik — APPROVED** [review]
One small nit, but no reason to stop merge. I have no further remakrs.
/unhold
